### PR TITLE
Minor Autolathe Refactor | Material Dupe-B-Gone

### LINKED ISSUE #80755: FREE MATS!!!
## Issue Summary

You can get literally infinite materials as long as you have one sheet of the original (including bananium thanks to Airhorns).

## Reproduction:

Step 1: Upgrade any protolathe with sufficiently better servos.
Step 2: Upgrade an autolathe with sufficiently better servos.
Step 3: Make an item from the protolathe and get a discount on the material costs.
Step 4: Get the undiscounted material costs back by putting it into an autolathe.
Step 5: Repeat for infinite profit.

Optional Step 6: Use a recycler instead since it can handle larger loads.

https://github.com/tgstation/tgstation/assets/106436013/03be8ca3-0a6c-473f-b78d-55f76e9417b8

Materials Out Vs Materials In:

image

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Refactors underlying autolathe code, mostly about how it prints items
Items are now printed sequentially
Items now have their custom materials set to the the materials actually used to create them
Items are now printed based on their design construction time instead of a default 3.2 seconds per item always
Closes https://github.com/tgstation/tgstation/issues/80755

## Changelog
:cl:
fix: Autolathes no longer allow you to duplicate materials at higher levels of stock parts
qol: Autolathes now show name instead of typepath when selecting a custom material
qol: Autolathe now print out items one by one instead of waiting for all of them to print at once
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
> Items now have their custom materials set to the the materials actually used to create them

You need to apply this solution to all items printed from techfabs as well because their cost efficiency goes below 1. 

You can still print an discount item from techfab and throw it in the recycler and get profits

**Comment by ZephyrTFA** [comment]
> > Items now have their custom materials set to the the materials actually used to create them
> 
> You need to apply this solution to all items printed from techfabs(circuit printer, module duplicator, anything that makes stuff)as well because their cost efficiency goes below 1.
> 
> You can still print an discount item from techfab and throw it in the recycler/feed it to the autolathe and get profits

out of scope, will do it in probably the next week

**Comment by SyncIt21** [comment]
> out of scope, will do it probably the next week in a different PR

Then you need to unlink the issue as your PR will not close it. The issue reproduction steps states this

> Make an item from the protolathe and get a discount on the material costs.

The problem is the upgraded "protolathe" creates a pick axe using discounted costs not the autolathe. 

Autolathe has an initial cost multiplier of 1.6 and after upgrading it becomes 1 so there are no profits from recycling.
Protolathes however has an initial cost multiplier of 1 and it can become as low as 0.25 after upgrading(but the printed item still contains the same amount of material) 

Techfabs are the real problem not autolathes

File: code/modules/research/machinery/_production.dm
Comment (SyncIt21) [inline code/modules/research/machinery/_production.dm:294]: Autodoc with params. Maybe make it PRIVATE(TRUE) if not used anywhere but that's your call

File: code/modules/research/machinery/_production.dm
Comment (SyncIt21) [inline code/modules/research/machinery/_production.dm:306]: This proc is just 2 lines. You could easily move it's code to where it's required or autodoc it, maybe make it PRIVATE(TRUE) but that's your call

File: code/modules/research/machinery/_production.dm
Comment (SyncIt21) [inline code/modules/research/machinery/_production.dm:240]: This list is not required after you follow the below reviews

File: code/game/machinery/autolathe.dm
Comment (SyncIt21) [inline code/game/machinery/autolathe.dm:254]: This proc is just 2 lines. You could easily move it's code to where it's required or autodoc it, maybe make it PRIVATE(TRUE) but that's your call

File: code/game/machinery/autolathe.dm
Comment (SyncIt21) [inline code/game/machinery/autolathe.dm:246]: Autodoc with params

File: code/datums/components/material/material_container.dm
Comment (SyncIt21) [inline code/datums/components/material/material_container.dm:586]: You needed this only for debugging right? If so please remove it

File: code/game/machinery/autolathe.dm
Comment (SyncIt21) [inline code/game/machinery/autolathe.dm:237]: Autodoc with params, maybe make it PRIVATE(TRUE) but that's your choice

File: code/game/machinery/autolathe.dm
Comment (SyncIt21) [inline code/game/machinery/autolathe.dm:220]: Use the 2nd and 3rd params instead of creating a new list called total_materials and doing all the scaling yourself.

It's just codebloat for no reason.

**Review by SyncIt21 — CHANGES_REQUESTED** [review]
Reviews. Instead of creating a new list which contains the total cost of the design with the print quantity and the machine efficiency applied and then passing that to use_materials()
Just use the 2nd and 3rd params of use_materials() to do that scaling for you. No need to do the work yourself

**Comment by SyncIt21** [comment]
Also remove setting custom materials for items printed from the autolathe.

This is because the autolathe can print items like the foam riot ammo box which contains 40 pellets each with it's own iron content.

The cost of items for the autolathe has been set in such a way that recycling these items snd their contents recursively will yield the exact same amount of material like it's cost.

The most accurate solution is to iterate recursively over all the contents of the printed object and set their custom materials such that their final sum becomes equal to the design cost but that's too complicated so for now you can remove settings custom materials for the autolathe

**Review by ArcaneMusic — APPROVED** [review]
LGTM

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
After getting some sleep my fresh eyes caught these two things

**Review by ArcaneMusic — APPROVED** [review]

**Comment by githubuser4141** [comment]
What the heck, this does an unmentioned thing, it prevents being used on the aux base because there's no APC on the aux base.
Why doesn't TGstation maintainers require all of the truth to be put into the changelog? They let favored people just make up PRs that do all sorts of things that aren't mentioned at all.

**Comment by ZephyrTFA** [comment]
> What the heck, this does an unmentioned thing, it prevents being used on the aux base because there's no APC on the aux base. Why doesn't TGstation maintainers require all of the truth to be put into the changelog? They let favored people just make up PRs that do all sorts of things that aren't mentioned at all.

Pretty obvious alt so I'll be blocking you from the repo.
But before you go, the autolathe was always intended to require power to print items. Its just that the previous code didnt actually check that it was able to use the power before printing.

**Pull request merged**
